external-crates/move/move-compiler/src/sui_mode/typing.rs

```diff
@@ -808,3 +920,124 @@ fn exp(context: &mut Context, e: &T::Exp) {
         _ => (),
     }
 }
+
+fn check_event_emit(context: &mut Context, loc: Loc, mcall: &ModuleCall) {
+    let current_module = context.current_module();
+    let ModuleCall {
+        module,
+        name,
+        type_arguments,
+        ..
+    } = mcall;
+    let Some(first_ty) = type_arguments.first() else {
+        // invalid arity
+        debug_assert!(false, "ICE arity should have been expanded for errors");
+        return;
+    };
+    let is_defined_in_current_module = matches!(first_ty.value.type_name(), Some(sp!(_, TypeName_::ModuleType(m, _))) if m == current_module);
+    if !is_defined_in_current_module {
+        let msg = format!(
+            "Invalid event. The function '{}::{}' must be called with a type defined in the current module, '{}'",
+            module, name, current_module
+        );
+        let ty_msg = format!(
+            "The type {} is not declared in the current module, '{}'",
+            error_format(first_ty, &Subst::empty()),
+            current_module
+        );
+        context.env.add_diag(diag!(
+            EVENT_EMIT_CALL_DIAG,
+            (loc, msg),
+            (first_ty.loc, ty_msg)
+        ));
+    }
+}
+
+fn check_private_transfer(context: &mut Context, loc: Loc, mcall: &ModuleCall) {
+    let ModuleCall {
+        module,
+        name,
+        type_arguments,
+        ..
+    } = mcall;
+    let current_module = context.current_module();
+    if current_module
+        .value
+        .is(SUI_ADDR_NAME, TRANSFER_FUNCTION_NAME)
+    {
+        // inside the transfer module, so no private transfer rules
+        return;
+    }
+    let Some(first_ty) = type_arguments.first() else {
+        // invalid arity
+        debug_assert!(false, "ICE arity should have been expanded for errors");
+        return;
+    };
+    let (in_current_module, first_ty_tn) = match first_ty.value.type_name() {
+        Some(sp!(_, TypeName_::Multiple(_))) | Some(sp!(_, TypeName_::Builtin(_))) | None => {
+            (false, None)
+        }
+        Some(sp!(_, TypeName_::ModuleType(m, n))) => (m == current_module, Some((m, n))),
+    };
+    if !in_current_module {
+        let mut msg = format!(
+            "Invalid private transfer. \
+            The function '{}::{}' is restricted to being called in the object's module",
+            module, name,
+        );
+        if let Some((first_ty_module, _)) = &first_ty_tn {
+            msg = format!("{}, '{}'", msg, first_ty_module);
+        };
+        let ty_msg = format!(
+            "The type {} is not declared in the current module, '{}'",
+            error_format(first_ty, &Subst::empty()),
+            current_module
+        );
+        let mut diag = diag!(
+            PRIVATE_TRANSFER_CALL_DIAG,
+            (loc, msg),
+            (first_ty.loc, ty_msg)
+        );
+        if first_ty
+            .value
+            .has_ability_(Ability_::Store)
+            .is_some_and(|b| b)
```

Comment:
```suggestion
            .is_some()
```

Reply:
The suggestion would change the meaning, since `is_some_and(|b| b)` will evaluate to `false` for `Some(false)`

Reply:
😭